### PR TITLE
[TACHYON-619] Improve logic of legalCascadingPlan

### PR DESCRIPTION
**Pull request by uronce-cc**
follow up PR https://github.com/amplab/tachyon/pull/1079

### CODE REVIEW

**Comment by uronce-cc** [comment]
@calvinjia

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 1]
```diff
@@ -42,46 +40,57 @@ public static boolean legalCascadingPlan(long bytesToBeAvailable, EvictionPlan p
...
-    int currentTierAlias = tierAliases.get(0);
-    // first tier to free space from needs to have bytesToBeAvailable after eviction
-    boolean isLegal = bytesToBeAvailableInTier.get(currentTierAlias) >= bytesToBeAvailable;
-    for (int nextTierAlias : tierAliases.subList(1, tierAliases.size())) {
-      if (!isLegal) {
-        break;
+    for (StorageDir dir : spaceInfoInDir.keySet()) {
+      Pair<Long, Long> spaceInfo = spaceInfoInDir.get(dir);
+      if (spaceInfo.getFirst() < spaceInfo.getSecond()
+          || spaceInfo.getFirst() < bytesToBeAvailable) {
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: The second condition (actually should be `spaceInfo.getFirst() - spaceInfo.getSecond()` only needs to be true for one dir right?

Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: good catch :+1:

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 2]
```diff
@@ -42,46 +40,68 @@ public static boolean legalCascadingPlan(long bytesToBeAvailable, EvictionPlan p
...
 
-    // upper to lower tier
-    Collections.sort(tierAliases);
-    int currentTierAlias = tierAliases.get(0);
-    // first tier to free space from needs to have bytesToBeAvailable after eviction
-    boolean isLegal = bytesToBeAvailableInTier.get(currentTierAlias) >= bytesToBeAvailable;
-    for (int nextTierAlias : tierAliases.subList(1, tierAliases.size())) {
-      if (!isLegal) {
-        break;
+    StorageDir firstTierDir = null;
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: Would it be possible to avoid assuming only one dir will be affected by eviction in the first tier (for example you could just make sure the max of the free space of dirs in the first tier is > bytesToBeAvailable)?

Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: make sense.

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 3]
```diff
@@ -42,46 +40,71 @@ public static boolean legalCascadingPlan(long bytesToBeAvailable, EvictionPlan p
       BlockMetadataManager metaManager) throws IOException {
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: NIT: word legal is probably not very proper. How about `isCascadingPlanValid`?

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 4]
```diff
@@ -16,72 +16,95 @@
...
+    // after the plan taking action
+    Map<StorageDir, Pair<Long, Long>> spaceInfoInDir = new HashMap<StorageDir, Pair<Long, Long>>();
+
+    for (long blockId : plan.toEvict()) {
+      BlockMeta block = metaManager.getBlockMeta(blockId);
+      StorageDir dir = block.getParentDir();
+      if (spaceInfoInDir.containsKey(dir)) {
+        Pair<Long, Long> spaceInfo = spaceInfoInDir.get(dir);
+        spaceInfo.setFirst(spaceInfo.getFirst() + block.getBlockSize());
+        spaceInfoInDir.put(dir, spaceInfo);
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: I don't think line 54 is necessary as `spaceInfo` is the reference to the `Pair`, could you doublecheck?

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 5]
```diff
@@ -16,72 +16,95 @@
...
-      if (bytesToBeAvailableInTier.containsKey(tierAlias)) {
-        bytesToBeAvailableInTier
-            .put(tierAlias, bytesToBeAvailableInTier.get(tierAlias) + blockSize);
+      StorageDir srcDir = block.getParentDir();
+      StorageDir destDir = metaManager.getDir(move.getSecond());
+
+      if (spaceInfoInDir.containsKey(srcDir)) {
+        Pair<Long, Long> spaceInfo = spaceInfoInDir.get(srcDir);
+        spaceInfo.setFirst(spaceInfo.getFirst() + blockSize);
+        spaceInfoInDir.put(srcDir, spaceInfo);
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: Same here, not sure line 71 is really useful

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 6]
```diff
@@ -16,72 +16,92 @@
...
-            .put(tierAlias, metaManager.getAvailableBytes(blockDir) + blockSize);
+        spaceInfoInDir.put(destDir, new Pair<Long, Long>(destDir.getAvailableBytes(), blockSize));
+      }
+    }
+
+    int firstTierAlias = Integer.MAX_VALUE;
+    for (StorageDir dir : spaceInfoInDir.keySet()) {
+      firstTierAlias = Math.min(firstTierAlias, dir.getParentTier().getTierAlias());
+    }
+    long firstTierDirMaxSpace = Long.MIN_VALUE;
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: variable name `firstTierDirMaxSpace` is very confusing. Please either name it more descriptive if possible, or comment this variable.

Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: what about `maxDirSpaceInFirstTier` ?

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 7]
```diff
@@ -16,72 +16,92 @@
...
 import tachyon.Pair;
 import tachyon.worker.block.BlockMetadataManager;
 import tachyon.worker.block.BlockStoreLocation;
 import tachyon.worker.block.meta.BlockMeta;
+import tachyon.worker.block.meta.StorageDir;
 
 public class EvictorUtils {
   /**
-   * Checks whether the plan is legal for an evictor with cascading eviction feature like
+   * Checks whether the plan is valid for an evictor with cascading eviction feature like
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: I think this javadoc is not clear enough about what this method is doing. At least I didn't get it at the first time. How about changing it to: 

Checks whether a given plan with cascading eviction is valid. [and please explain what is cascading---I still don't have a good sense of this definition]. A plan may become invalid on cascading eviction if the total size of blocks to move to a lower tier is larger than the total amount of available space of this lower tier even after its blocks evicted.

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 8]
```diff
@@ -16,72 +16,92 @@
...
+        spaceInfo.setSecond(spaceInfo.getSecond() + blockSize);
       } else {
-        tierAliases.add(tierAlias);
-        bytesToBeAvailableInTier
-            .put(tierAlias, metaManager.getAvailableBytes(blockDir) + blockSize);
+        spaceInfoInDir.put(destDir, new Pair<Long, Long>(destDir.getAvailableBytes(), blockSize));
+      }
+    }
+
+    int firstTierAlias = Integer.MAX_VALUE;
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: Question: what do you mean by first tier? top tier? bottom tier?

Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: the most top tier of all tiers where blocks in the eviction plan reside in.

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 9]
```diff
@@ -16,72 +16,92 @@
...
+    }
+    long maxSpace = Long.MIN_VALUE; // maximum bytes to be available in a dir in the first tier
+    for (StorageDir dir : spaceInfoInDir.keySet()) {
+      if (dir.getParentTier().getTierAlias() == firstTierAlias) {
+        Pair<Long, Long> space = spaceInfoInDir.get(dir);
+        maxSpace = Math.max(maxSpace, space.getFirst() - space.getSecond());
       }
     }
+    if (maxSpace < bytesToBeAvailable) {
+      return false;
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: Comment on why this is an invalid plan

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 10]
```diff
@@ -16,72 +16,92 @@
...
-    int currentTierAlias = tierAliases.get(0);
-    // first tier to free space from needs to have bytesToBeAvailable after eviction
-    boolean isLegal = bytesToBeAvailableInTier.get(currentTierAlias) >= bytesToBeAvailable;
-    for (int nextTierAlias : tierAliases.subList(1, tierAliases.size())) {
-      if (!isLegal) {
-        break;
+    for (StorageDir dir : spaceInfoInDir.keySet()) {
+      Pair<Long, Long> spaceInfo = spaceInfoInDir.get(dir);
+      if (spaceInfo.getFirst() < spaceInfo.getSecond()) {
+        return false;
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: Comment on why this is an invalid plan

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 11]
```diff
@@ -16,72 +16,92 @@
...
    * enough space in lower tier to move in blocks from upper tier.
    *
    * @param bytesToBeAvailable requested bytes to be available after eviction
-   * @param plan the eviction plan
+   * @param plan the eviction plan, should not be empty
    * @param metaManager the meta data manager
    * @return true if the above requirements are satisfied, otherwise false.
    */
-  public static boolean legalCascadingPlan(long bytesToBeAvailable, EvictionPlan plan,
+  public static boolean validCascadingPlan(long bytesToBeAvailable, EvictionPlan plan,
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: Also, please unit-test this method --- let's move that in another PR

File: servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java [hunk 12]
```diff
@@ -16,72 +16,101 @@
...
 
-    // upper to lower tier
-    Collections.sort(tierAliases);
-    int currentTierAlias = tierAliases.get(0);
-    // first tier to free space from needs to have bytesToBeAvailable after eviction
-    boolean isLegal = bytesToBeAvailableInTier.get(currentTierAlias) >= bytesToBeAvailable;
-    for (int nextTierAlias : tierAliases.subList(1, tierAliases.size())) {
-      if (!isLegal) {
-        break;
+    // the most top tier among all tiers where blocks in the plan reside in
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: typo: "most top" -> "top". Unnecessary to add "most" before "top"

Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java]: resolved.

File: servers/src/main/java/tachyon/worker/block/evictor/LRUEvictor.java [hunk 13]
```diff
@@ -209,7 +211,7 @@ public EvictionPlan freeSpace(long bytesToBeAvailable, BlockStoreLocation locati
       }
     }
 
-    return EvictorUtils.legalCascadingPlan(bytesToBeAvailable, plan, mMetaManager) ? plan : null;
+    return EvictorUtils.validCascadingPlan(bytesToBeAvailable, plan, mMetaManager) ? plan : null;
```
Comment (apc999) [inline servers/src/main/java/tachyon/worker/block/evictor/LRUEvictor.java]: Question: is calling `EvictorUtils.validCascadingPlan` here just to verify the plan is correct, or it is required for the return value? If it is not required, we shall not have the verification on the critical code path---during eviction heavy lock is hold.

**Comment by apc999** [comment]
One question left to help me understand the code better. Besides, it looks good to me now.

**Comment by haoyuan** [comment]
We can’t automatically merge this pull request.

File: servers/src/main/java/tachyon/worker/block/evictor/LRUEvictor.java (hunk 13 above)
Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/LRUEvictor.java]: @apc999 just to verify the plan is correct.

Comment (uronce-cc) [inline servers/src/main/java/tachyon/worker/block/evictor/LRUEvictor.java]: @apc999 I'll remove both this validation and the verification of block existence above in PR: https://github.com/amplab/tachyon/pull/1097

**Comment by uronce-cc** [comment]
@haoyuan merged.

**Comment by haoyuan** [comment]
We can’t automatically merge this pull request.

**Comment by uronce-cc** [comment]
@haoyuan conflict resolved.

**Comment by apc999** [comment]
+1 LGTM. 

Note that, one PR needed to cleanup the verification of valid plan on the critical path of freeSpaceWithView.

**Comment by uronce-cc** [comment]
@apc999 for communication efficiency, I'll do the cleanup in this PR with one commit since the code changes are small.

**Comment by calvinjia** [comment]
+1 LGTM

**Comment by apc999** [comment]
+1 LGTM
